### PR TITLE
[BEAM-14347] Add some benchmarks for generic registration

### PR DESCRIPTION
**Pull request by damccorm**
This adds some basic benchmarks to prove that the generic registration works as well as the generated version. This isn't really perfect - it largely looks at pieces of the overall flow and doesn't provide a comprehensive full run of the DoFn, and it misses some cases (e.g. iters/emitters). This is because:

1) We can't have these benchmarks in the exec package because including the register package there induces a circular dependency and go test throws.
2) Much of the stuff we'd like to be able to easily test is private in exec. An example is that makeEmit is private, and its pretty challenging to test that specific functionality.

I didn't try _too_ hard to work around those constraints because I expect that once we've built up a decent history of load test runs that will actually be a better way to show that generic registration is just as good for more "real" use cases.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: sdks/go/pkg/beam/register/register_test.go [hunk 1]
```diff
@@ -672,3 +675,133 @@ func (fn *PartialCombiner2) AddInput(i int, c CustomType) int {
...
+	return f.A + b.val
+}
+
+func MakeMultiEdge(f *graph.DoFn) graph.MultiEdge {
+	return graph.MultiEdge{
+		DoFn: f,
+	}
+}
+
+type callerCustomTypeГint struct {
```
Comment (riteshghorse) [inline sdks/go/pkg/beam/register/register_test.go:696]:
```suggestion
type callerCustomTypeГint struct {
```
is this character intentional?

**Review by riteshghorse — APPROVED** [review]

File: sdks/go/pkg/beam/register/register_test.go (hunk 1 above)
Comment (riteshghorse) [reply sdks/go/pkg/beam/register/register_test.go:696]: okay.. interesting!

File: sdks/go/pkg/beam/register/register_test.go [hunk 2]
```diff
@@ -672,3 +675,133 @@ func (fn *PartialCombiner2) AddInput(i int, c CustomType) int {
...
+// BenchmarkMethodCalls/EncodeMultiEdge_Unoptimized-16               1000000	      	1063 ns/op
+//
+// BenchmarkMethodCalls/MakeFunc_GenericRegistration-16             16266259	        72.07 ns/op
+// BenchmarkMethodCalls/NewFn_GenericRegistration-16                 1000000	      	1108 ns/op
+// BenchmarkMethodCalls/EncodeMultiEdge_GenericRegistration-16       1000000	      	1052 ns/op
+//
+// BenchmarkMethodCalls/MakeFunc_GeneratedShims#01-16               16400914	        69.17 ns/op
+// BenchmarkMethodCalls/NewFn_GeneratedShims#01-16                   1000000	      	1099 ns/op
+// BenchmarkMethodCalls/EncodeMultiEdge_GeneratedShims#01-16         1000000	      	1071 ns/op
+func BenchmarkMethodCalls(b *testing.B) {
```
[the remark this answers is not in the document]
Comment (damccorm) [reply sdks/go/pkg/beam/register/register_test.go:763]: You're right - that was a silly oversight, I'd kinda lost track of some of my initial goals as I was writing these. As I was fixing it, I actually found an issue with the code we're generating (so thanks for calling that out!) that was causing it to not quite match the code generator, though it was still close.

I fixed both issues

[the remark this answers is not in the document]
Comment (lostluck) [reply sdks/go/pkg/beam/register/register_test.go:763]: And thus, revealing why I felt so strongly about this sort of thing :P. Thanks!

[the remark this answers is not in the document]
Comment (damccorm) [reply sdks/go/pkg/beam/register/register_test.go:763]: Yep - good instincts!

**Review by lostluck — APPROVED** [review]
LGTM! I didn't intend for the benchmark to catch the generation error, but I'm very glad it did. This is lovely. Thanks!

**Pull request merged**